subworkflows/local/align_cellrangermulti.nf

```diff
@@ -0,0 +1,226 @@
+//
+// Include modules
+//
+include { CELLRANGER_MKGTF                  } from "../../modules/nf-core/cellranger/mkgtf/main.nf"
+include { CELLRANGER_MKREF                  } from "../../modules/nf-core/cellranger/mkref/main.nf"
+include { CELLRANGER_MKVDJREF               } from "../../modules/nf-core/cellranger/mkvdjref/main.nf"
+include { CELLRANGER_MULTI                  } from "../../modules/nf-core/cellranger/multi/main.nf"
+include { PARSE_CELLRANGERMULTI_SAMPLESHEET } from "../../modules/local/parse_cellrangermulti_samplesheet.nf"
+
+// Define workflow to subset and index a genome region fasta file
+workflow CELLRANGER_MULTI_ALIGN {
+    take:
+        ch_fasta
+        ch_gtf
+        ch_fastq
+        cellranger_gex_index
+        cellranger_vdj_index
+        ch_multi_samplesheet
+
+    main:
+        ch_versions    = Channel.empty()
+
+        //
+        // TODO: Include checkers for cellranger multi parameter combinations. For example, when VDJ data is given, require VDJ ref. If FFPE, require frna probe sets, etc.
+        //
+
+        // since we merged all data as a meta, now we have a channel per sample, which
+        // every item is a meta map for each data-type
+        // now we can split it back for passing as input to the module
+        ch_fastq
+        .flatten()
+        .map{ meta ->
+            def meta_clone = meta.clone()
+            def data_dict  = meta_clone.find{ it.key == "${meta_clone.feature_type}" }
+            fastqs = data_dict?.value
+            meta_clone.remove( data_dict?.key )
+            [ meta_clone, fastqs ]
+        }
+        .branch {
+            meta, fastq ->
+                gex: meta.feature_type == "gex"
+                    return [ meta, fastq ]
+                vdj: meta.feature_type == "vdj"
+                    return [ meta, fastq ]
+                ab: meta.feature_type == "ab"
+                    return [ meta, fastq ]
+                beam: meta.feature_type == "beam"
+                    return [ meta, fastq ]
+                crispr: meta.feature_type == "crispr"
+                    return [ meta, fastq ]
+                cmo: meta.feature_type == "cmo"
+                    return [ meta, fastq ]
+        }
+        .set { ch_grouped_fastq }
+
+        // Assign other cellranger reference files
+        ch_gex_frna_probeset      = params.gex_frna_probe_set            ? file(params.gex_frna_probe_set)            : []
+        ch_gex_target_panel       = params.gex_target_panel              ? file(params.gex_target_panel)              : []
+        ch_gex_cmo_set            = params.gex_cmo_set                   ? file(params.gex_cmo_set)                   : []
+        ch_gex_barcodes           = params.gex_barcode_sample_assignment ? file(params.gex_barcode_sample_assignment) : []
```

Comment:
I don't think this is used anywhere? Or am I missing something... 
It would make sense because the barcode assignments come from our second samplesheet.

Reply:
they are used in the `cellranger multi` script, module and template, to generate the config file:

https://github.com/nf-core/scrnaseq/blob/f4304ad0b187af801b2682f0e4870b2e0ed9f3f4/modules/nf-core/cellranger/multi/templates/cellranger_multi.py

Reply:
But, indeed I do not use in the testings the parameter `--gex_barcode_sample_assignment`.
Thus, I think this is the only one that might be removed.

The other ones, I understand the would be needed, as in order they would possibly mean:

* the reference/sequence information of the probes (not the sample assignment)
* the reference of the ab targets (not the sample assignment)
* the reference/sequence information of the CMO probes (not the sample assignment)
* 
What do you think? I can try removing completely the `ch_gex_barcodes` here and see if it works. Which, as you said, this is the one coming from the second samplesheet.

Reply:
Sorry, I was referring to ch_gex_barcodes specifically... 

I can see where it is used now, but I am still confused. Shouldn't the "barcode sample assignment" be generated from that samplesheet we provide like `ch_cmo_barcode_csv` and `ch_frna_sample_csv`?

Reply:
I think I need to do some more research about what these three files are
```
    path cmo_reference         , stageAs: "references/cmo/*"
    path cmo_barcodes          , stageAs: "references/cmo/barcodes/*"
    path cmo_barcode_assignment, stageAs: "references/cmo/sample_barcode_assignment/*"
```

I understand the purpose of the assignment and the reference, but not the third one.

Reply:
Yes, I think the same. I am going to do a testing / researching on this `cmo_barcode_assignment` now to check if that is the case. But I agree that it is probably what is coming from our second samplesheet.

I will come back when I figure something out.

Reply:
Hi @grst ,
I checked it again and you are correct.
Based on the documentation, for the CMO, we only need the information that is currently given via the 2nd samplesheet:

https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/running-pipelines/cr-3p-multi

Currently I removed the parameter because it really does not seem necessary.

As we are doing for BEAM, I can add a note, `currently not supported`.

Reply:
I still don't quite get it... but to me it seems `sample_barcode_assignment` is a redundant way of specifying sample<->cmo mappings:
![image](https://github.com/nf-core/scrnaseq/assets/7051479/8674dac0-7b3b-49ee-9a77-1fc48d829733)


I would just remove support for `sample_barcode_assignment` for now. If this is required for any usecase, someone will request it eventually.

Reply:
Never mind, I finally found the right documentation: 

> The barcode-sample-assignment option in the [gene-expression] section of the multi config CSV allows users to provide a file that manually specifies the barcodes for each sample. It will override Cell Ranger's default cell calling and tag calling steps, and may be useful in cases where data with microfluidic failures can be partially rescued. This feature allows users to import custom tag calling done via 3rd party tools as well 

Seems there are some rare usecases for this indeed. Let's keep it then and I'll add the decription to the schema params.

Reply:
Okay. I will add it again ( had removed in the latest commit ).
I will mark this resolved once I bring it back.